Sound Management

cothon01 opened the pull request:
Welcome... to the _future_ of AUDIO MANAGEMENT!

I created a sound management system that facilitates the playing of sound clips (One shot, not looping). Why this is useful:
-> Removes inspector bloat: Your gameobjects don't need to store their own AudioSource components to play sounds
-> Easy Access: Uses a singleton (SFXManager) that can be used by any script with a single line of code
-> Uninterrupted Sound Playing: AudioSources are like a resource. You need more than one to play multiple sounds simultaneously. SFXManager, manages a bunch of them and ensures that no sound is interrupted (unless you play a ridiculous number of simultaneous sounds)
-> Supports 3D sound: SFXManager moves the AudioSource to a target point in case you need 3d sound.

I also included an AudioMixer with two sound channels (BGM and SFX). And a test scene that you can test this system with.

Bonus: I added SceneOnlyEdit.cs which is a modified version of the ReadOnly property drawer. By using the [SceneOnlyEdit] attribute, you mark an inspector variable so that it cannot be edited during Play Mode. But you can edit its value when the game is not running.

cothon01 committed:
Game Audio Management 1

+ Added an audio mixer with two channels (BGM for music, and SFX for sound effects)
+ Create SFXClip data object to help with referencing multiple sound clips with a single data object.

cothon01 committed:
Game Audio Management 2

+ Added Singleton implementation
+ Singleton persistent flag to control whether object is destroyed on scene load or not.

cothon01 committed:
Game Audio Management 3

+ SFXManager script singleton that creates a collection of audiosources
+ A test scene for the SFXManager script

cothon01 committed:
Game Audio Management 3

+Added function in SFXManager to find unoccupied AudioSource component.
+ Added function that plays sound on AudioSource component
+ Added public functions for playing 2d and 3d sounds
+ Overloads for playing specific AudioClips, or SFXClip (for random clip)

cothon01 committed:
Game Audio Management 4

+Fixed some bugs in SFXManager
+Added debug GUI for SFXManager
+Sound testing scene (click to play sound)

cothon01 committed:
Scene Only Edit Attribute

I created a property attribute that is identical to ReadOnly.cs
The only difference is that it disables GUI during PlayMode.
This still allows devs to change values in the inspector, but not while the game is running.

Helcostr commented:
oooo fancy.

Sneakydeaky reviewed (approved):
Wow, That's Awesome! Excellent job.

Helcostr commented:
There seems to be an issue when the source amount is too low, currently playing sounds get overwritten. Is this intentional? Are you going to dynamically allocate more AudioSources?

Helcostr commented on Assets/Scripts/Audio/SFXManager.cs:
```diff
@@ -0,0 +1,186 @@
+                break;
+            }
+        }
+        return source;
```
This will return sources[0] if all sources are preoccupied.

cothon01 replied:
Just a minor setback ^-^' ...

Helcostr reviewed (changes requested):
Please check and approve/change the commented line.

cothon01 committed:
Game Audio Management 5

Returns random audio source when you run out and prints warning for the devs, so that they may increase the audio sources.

Helcostr reviewed (approved):
No issues.
